File: src/core/platforms.ts
import fs from "node:fs";
import path from "node:path";
import * as TOML from "@iarna/toml";
import { expandHome } from "./config.js";

export type PlatformInstallSpec = {
  supported?: boolean;
  build_path?: string;
  install_path_env?: string;
  install_path_default?: string;
  install_root_env?: string;
  install_root_default?: string;
  install_path_suffix?: string;
  scope?: "user" | "project";
  format?: "codex_toml" | "json_map";
  root_key?: string;
};

export type PlatformAuthSpec = {
  mode?: string;
  instructions?: string[];
};

export type PlatformDefinition = {
  id: string;
  display_name: string;
  skill?: PlatformInstallSpec;
  mcp?: PlatformInstallSpec;
  manifest?: PlatformInstallSpec;
  auth?: PlatformAuthSpec;
  __file__: string;
};

export function formatTemplate(template: string, context: Record<string, string>): string {
  return template.replace(/\{([^}]+)\}/g, (match, key: string) => {
    if (!(key in context)) throw new Error(`Missing template variable '${key}' in value: ${template}`);
    return context[key] ?? match;
  });
}

export function expandPath(value: string, context: Record<string, string>): string {
  return expandHome(formatTemplate(value, context));
}

export function contextForPaths(skillName: string, repoPath: string | null): Record<string, string> {
  return { skill_name: skillName, repo_path: repoPath ?? "" };
}

export function loadPlatforms(platformsDir: string): Record<string, PlatformDefinition> {
  const platforms: Record<string, PlatformDefinition> = {};
  for (const entry of fs.readdirSync(platformsDir)) {
    if (!entry.endsWith(".toml")) continue;
    const filePath = path.join(platformsDir, entry);
    const raw = fs.readFileSync(filePath, "utf8");
    const parsed = TOML.parse(raw) as Omit<PlatformDefinition, "__file__">;
    if (!parsed.id) throw new Error(`Platform file missing id: ${filePath}`);
    if (parsed.id !== path.basename(entry, ".toml")) {
      throw new Error(`Platform id '${parsed.id}' must match filename: ${filePath}`);
    }
    platforms[parsed.id] = { ...parsed, __file__: filePath };
  }
  if (Object.keys(platforms).length === 0) {
    throw new Error(`No platform definitions found in ${platformsDir}`);
  }
  return platforms;
}

function normalizePlatformId(value: string): string {
  return value.toLowerCase().replace(/[^a-z0-9]/g, "");
}

function levenshteinDistance(a: string, b: string): number {
  if (a === b) return 0;
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;

  const previous = Array.from({ length: b.length + 1 }, (_, index) => index);

  for (let i = 1; i <= a.length; i += 1) {
    let diagonal = previous[0]!;
    previous[0] = i;
    for (let j = 1; j <= b.length; j += 1) {
      const nextDiagonal = previous[j]!;
      const substitutionCost = a[i - 1] === b[j - 1] ? 0 : 1;
      previous[j] = Math.min(
        previous[j]! + 1,
        previous[j - 1]! + 1,
        diagonal + substitutionCost,
      );
      diagonal = nextDiagonal;
    }
  }

  return previous[b.length]!;
}

function suggestPlatformId(platformIds: string[], input: string): string | null {
  const normalizedInput = normalizePlatformId(input);
  if (!normalizedInput) return null;

  let bestPrefixMatch: string | null = null;
  let bestDistanceMatch: { id: string; distance: number } | null = null;

  for (const platformId of platformIds) {
    const normalizedPlatformId = normalizePlatformId(platformId);
    if (
      normalizedPlatformId.startsWith(normalizedInput) ||
      normalizedInput.startsWith(normalizedPlatformId) ||
      normalizedPlatformId.includes(normalizedInput)
    ) {
      if (!bestPrefixMatch || platformId.length < bestPrefixMatch.length) bestPrefixMatch = platformId;
      continue;
    }

    const distance = levenshteinDistance(normalizedInput, normalizedPlatformId);
    if (!bestDistanceMatch || distance < bestDistanceMatch.distance) {
      bestDistanceMatch = { id: platformId, distance };
    }
  }

  if (bestPrefixMatch) return bestPrefixMatch;
  if (!bestDistanceMatch) return null;

  const maxDistance = normalizedInput.length <= 4 ? 1 : 2;
  return bestDistanceMatch.distance <= maxDistance ? bestDistanceMatch.id : null;
}

function unknownPlatformsMessage(platformIds: string[], missing: string[]): string {
  const base = `Unknown platform(s): ${missing.join(", ")}.`;
  if (missing.length === 1) {
    const suggestion = suggestPlatformId(platformIds, missing[0]!);
    if (suggestion) return `${base} Did you mean '${suggestion}'? Run 'dops platform list' for supported platforms.`;
    return `${base} Run 'dops platform list' for supported platforms.`;
  }

  const suggestions = missing
    .map((id) => {
      const suggestion = suggestPlatformId(platformIds, id);
      return suggestion ? `'${id}' -> '${suggestion}'` : null;
    })
    .filter((value): value is string => Boolean(value));

  if (suggestions.length > 0) {
    return `${base} Suggestions: ${suggestions.join(", ")}. Run 'dops platform list' for supported platforms.`;
  }

  return `${base} Run 'dops platform list' for supported platforms.`;
}

export function selectPlatforms(
  platforms: Record<string, PlatformDefinition>,
  selectedIds?: string[],
  capability?: "skill" | "mcp" | "manifest",
): PlatformDefinition[] {
  const orderedIds = selectedIds && selectedIds.length > 0 ? selectedIds : Object.keys(platforms);
  const missing = orderedIds.filter((id) => !platforms[id]);
  if (missing.length > 0) throw new Error(unknownPlatformsMessage(Object.keys(platforms), missing));
  return orderedIds
    .map((id) => platforms[id])
    .filter((p) => !capability || Boolean(p[capability]?.supported));
}

export function resolveInstallPath(spec: PlatformInstallSpec, context: Record<string, string>): string | null {
  if (spec.install_path_env && process.env[spec.install_path_env]) {
    return path.resolve(expandPath(process.env[spec.install_path_env]!, context));
  }
  if (spec.install_root_env || spec.install_root_default) {
    const rootValue = spec.install_root_env ? process.env[spec.install_root_env] || spec.install_root_default : spec.install_root_default;
    if (!rootValue) return null;
    const rootPath = path.resolve(expandPath(rootValue, context));
    return path.join(rootPath, formatTemplate(spec.install_path_suffix ?? "", context));
  }
  if (spec.install_path_default) {
    if (!context.repo_path && spec.install_path_default.includes("{repo_path}")) return null;
    return path.resolve(expandPath(spec.install_path_default, context));
  }
  return null;
}

export function authInstructions(platform: PlatformDefinition, context: Record<string, string>): string[] | null {
  if (platform.auth?.mode !== "interactive_handoff") return null;
  return (platform.auth.instructions ?? []).map((step) => formatTemplate(step, context));
}

File: dops/platforms.py
from __future__ import annotations

import os
import re
import tomllib
from dataclasses import dataclass
from pathlib import Path

from .config import expand_home
from .text_utils import levenshtein_distance


class InvalidPlatformDefinitionError(RuntimeError):
    def __init__(self, file_path: Path, details: str) -> None:
        self.file_path = file_path
        self.details = details
        super().__init__(
            f"DecisionOps platform definition is invalid: {file_path}: {details}. "
            "Refresh the skill bundle or provide a valid local skill checkout."
        )


@dataclass(slots=True)
class PlatformInstallSpec:
    supported: bool = False
    build_path: str | None = None
    install_path_env: str | None = None
    install_path_default: str | None = None
    install_root_env: str | None = None
    install_root_default: str | None = None
    install_path_suffix: str | None = None
    scope: str | None = None
    format: str | None = None
    root_key: str | None = None


@dataclass(slots=True)
class PlatformAuthSpec:
    mode: str | None = None
    instructions: list[str] | None = None


@dataclass(slots=True)
class PlatformDefinition:
    id: str
    display_name: str
    skill: PlatformInstallSpec | None
    mcp: PlatformInstallSpec | None
    manifest: PlatformInstallSpec | None
    auth: PlatformAuthSpec | None
    file_path: str


def format_template(template: str, context: dict[str, str]) -> str:
    def replace(match: re.Match[str]) -> str:
        key = match.group(1)
        if key not in context:
            raise RuntimeError(f"Missing template variable '{key}' in value: {template}")
        return context[key]

    return re.sub(r"\{([^}]+)\}", replace, template)


def expand_path(value: str, context: dict[str, str]) -> str:
    return expand_home(format_template(value, context))


def context_for_paths(skill_name: str, repo_path: str | None) -> dict[str, str]:
    return {"skill_name": skill_name, "repo_path": repo_path or ""}


def _install_spec(data: dict[str, object] | None) -> PlatformInstallSpec | None:
    if data is None:
        return None
    return PlatformInstallSpec(
        supported=bool(data.get("supported", False)),
        build_path=_optional_str(data.get("build_path")),
        install_path_env=_optional_str(data.get("install_path_env")),
        install_path_default=_optional_str(data.get("install_path_default")),
        install_root_env=_optional_str(data.get("install_root_env")),
        install_root_default=_optional_str(data.get("install_root_default")),
        install_path_suffix=_optional_str(data.get("install_path_suffix")),
        scope=_optional_str(data.get("scope")),
        format=_optional_str(data.get("format")),
        root_key=_optional_str(data.get("root_key")),
    )


def _auth_spec(data: dict[str, object] | None) -> PlatformAuthSpec | None:
    if data is None:
        return None
    instructions = data.get("instructions")
    return PlatformAuthSpec(
        mode=_optional_str(data.get("mode")),
        instructions=[str(item) for item in instructions] if isinstance(instructions, list) else None,
    )


def _optional_str(value: object) -> str | None:
    return str(value) if value is not None else None


def load_platforms(platforms_dir: str) -> dict[str, PlatformDefinition]:
    platforms: dict[str, PlatformDefinition] = {}
    for file_path in sorted(Path(platforms_dir).glob("*.toml")):
        try:
            parsed = tomllib.loads(file_path.read_text(encoding="utf8"))
        except tomllib.TOMLDecodeError as error:
            raise InvalidPlatformDefinitionError(file_path, str(error)) from error
        platform_id = str(parsed.get("id", ""))
        if not platform_id:
            raise RuntimeError(f"Platform file missing id: {file_path}")
        if platform_id != file_path.stem:
            raise RuntimeError(f"Platform id '{platform_id}' must match filename: {file_path}")
        platforms[platform_id] = PlatformDefinition(
            id=platform_id,
            display_name=str(parsed.get("display_name", platform_id)),
            skill=_install_spec(parsed.get("skill") if isinstance(parsed.get("skill"), dict) else None),
            mcp=_install_spec(parsed.get("mcp") if isinstance(parsed.get("mcp"), dict) else None),
            manifest=_install_spec(parsed.get("manifest") if isinstance(parsed.get("manifest"), dict) else None),
            auth=_auth_spec(parsed.get("auth") if isinstance(parsed.get("auth"), dict) else None),
            file_path=str(file_path),
        )
    if not platforms:
        raise RuntimeError(f"No platform definitions found in {platforms_dir}")
    return platforms


def _normalize_platform_id(value: str) -> str:
    return re.sub(r"[^a-z0-9]", "", value.lower())


def _suggest_platform_id(platform_ids: list[str], input_value: str) -> str | None:
    normalized_input = _normalize_platform_id(input_value)
    if not normalized_input:
        return None
    best_prefix_match: str | None = None
    best_distance_match: tuple[str, int] | None = None
    for platform_id in platform_ids:
        normalized_platform_id = _normalize_platform_id(platform_id)
        if (
            normalized_platform_id.startswith(normalized_input)
            or normalized_input.startswith(normalized_platform_id)
            or normalized_input in normalized_platform_id
        ):
            if best_prefix_match is None or len(platform_id) < len(best_prefix_match):
                best_prefix_match = platform_id
            continue
        distance = levenshtein_distance(normalized_input, normalized_platform_id)
        if best_distance_match is None or distance < best_distance_match[1]:
            best_distance_match = (platform_id, distance)
    if best_prefix_match:
        return best_prefix_match
    if best_distance_match is None:
        return None
    max_distance = 1 if len(normalized_input) <= 4 else 2
    return best_distance_match[0] if best_distance_match[1] <= max_distance else None


def _unknown_platforms_message(platform_ids: list[str], missing: list[str]) -> str:
    base = f"Unknown platform(s): {', '.join(missing)}."
    if len(missing) == 1:
        suggestion = _suggest_platform_id(platform_ids, missing[0])
        if suggestion:
            return f"{base} Did you mean '{suggestion}'? Run 'dops platform list' for supported platforms."
        return f"{base} Run 'dops platform list' for supported platforms."
    suggestions = []
    for item in missing:
        suggestion = _suggest_platform_id(platform_ids, item)
        if suggestion:
            suggestions.append(f"'{item}' -> '{suggestion}'")
    if suggestions:
        return f"{base} Suggestions: {', '.join(suggestions)}. Run 'dops platform list' for supported platforms."
    return f"{base} Run 'dops platform list' for supported platforms."


def select_platforms(
    platforms: dict[str, PlatformDefinition],
    selected_ids: list[str] | None = None,
    capability: str | None = None,
) -> list[PlatformDefinition]:
    ordered_ids = selected_ids if selected_ids else list(platforms.keys())
    missing = [platform_id for platform_id in ordered_ids if platform_id not in platforms]
    if missing:
        raise RuntimeError(_unknown_platforms_message(list(platforms.keys()), missing))
    selected = [platforms[platform_id] for platform_id in ordered_ids]
    if capability is None:
        return selected
    return [
        platform
        for platform in selected
        if getattr(platform, capability) is not None and bool(getattr(getattr(platform, capability), "supported", False))
    ]


def resolve_install_path(spec: PlatformInstallSpec, context: dict[str, str]) -> str | None:
    if spec.install_path_env and os.environ.get(spec.install_path_env):
        return str(Path(expand_path(os.environ[spec.install_path_env], context)).resolve())
    if spec.install_root_env or spec.install_root_default:
        root_value = (
            os.environ.get(spec.install_root_env, spec.install_root_default or "")
            if spec.install_root_env
            else spec.install_root_default
        )
        if not root_value:
            return None
        root_path = Path(expand_path(root_value, context)).resolve()
        return str(root_path / format_template(spec.install_path_suffix or "", context))
    if spec.install_path_default:
        if "{repo_path}" in spec.install_path_default and not context.get("repo_path"):
            return None
        return str(Path(expand_path(spec.install_path_default, context)).resolve())
    return None


def auth_instructions(platform: PlatformDefinition, context: dict[str, str]) -> list[str] | None:
    if platform.auth is None or platform.auth.mode != "interactive_handoff":
        return None
    return [format_template(step, context) for step in (platform.auth.instructions or [])]
